src/rsh/builtins/mod.rs

```diff
@@ -1,33 +1,93 @@
-use rsh::State;
 use std::path::PathBuf;
 use std::collections::HashMap;
-use std::slice::Iter;
 
-pub type Builtin = fn(State) -> State;
+use rsh::State;
+use rsh::utils;
+
+pub type Builtin = fn(&mut State) -> i32;
 
 pub fn load() -> HashMap<String, Builtin> {
     let mut h = HashMap::new();
 
-    h.insert("cd".to_string(), cd as fn(State) -> State);
-    h.insert("echo".to_string(), echo as fn(State) -> State);
-    // h.insert("ls".to_string(), ls as fn(State) -> State);
+    h.insert("cd".to_string(), cd as fn(&mut State) -> i32);
+    h.insert("ls".to_string(), ls as fn(&mut State) -> i32);
+    h.insert("echo".to_string(), echo as fn(&mut State) -> i32);
+    h.insert("set".to_string(), set as fn(&mut State) -> i32);
+    h.insert("unset".to_string(), unset as fn(&mut State) -> i32);
+    h.insert("get".to_string(), get as fn(&mut State) -> i32);
```

Comment:
Every function in rust is compiled to it's own special type and can be referenced by it's type signature. What I didn't realize and was able to find and correct because you asked this is the reason we were getting the weird type errors is because we didn't type the HashMap when we declared it so it assumed it was a `String` to` fn(&mut State) -> i32 { rsh::builtins::cd }` type which obviously is a type signature specific to the cd function (as it's the first item). So instead of correctly identifying the type of h at it's `let` binding I instead overrode the type inference at each builtin. Once I realized this was what I was actually doing I simply specified the type of h properly at it's let binding and now this code is much cleaner as of my latest commit. 

Thanks for asking this question made bad code good.